rsconnect/metadata.py

```diff
@@ -420,3 +430,203 @@ def resolve(self, server, app_id, app_mode):
+
+    def update_content_item(self, server, guid, content):
+        """
+        Update an existing item in the rebuilds for a given server
+        """
+        self._lock.acquire()
+        try:
+            old_content = self.get_content_item(server, guid)
+            if not old_content:
+                raise api.RSConnectException("Content not found: %s" % guid)
+
+            old_content.update(dict(
+                guid=guid,
+                bundle_id=content['bundle_id'],
+                title=content['title'],
+                name=content['name'],
+                app_mode=content['app_mode'],
+                content_url=content['content_url'],
+                dashboard_url=content['dashboard_url'],
+                created_time=content['created_time'],
+                last_deployed_time=content['last_deployed_time'],
+            ))
+            self.save()
+        finally:
+            self._lock.release()
+
+    def get_content_item(self, server, guid):
+        """
+        Get a content item from the rebuilds for a given server by guid
+        """
+        return self._data.get(server.url, {}).get('content', {}).get(guid)
+
+    def _cleanup_content_log_dir(self, server, guid):
+        """
+        Delete the local logs directory for a given content item.
+        """
+        logs_dir = self.get_rebuild_logs_dir(server, guid)
+        try:
+            shutil.rmtree(logs_dir)
+        except FileNotFoundError:
+            pass
+
+    def remove_content_item(self, server, guid, purge=False):
+        """
+        Remove a content item from the rebuilds for a given server by guid.
+        If purge is True, cleanup the log files on the local filesystem.
+        """
+        if purge:
+            self._cleanup_content_log_dir(server, guid)
+
+        self._lock.acquire()
+        try:
+            try:
+                self._data.get(server.url, {}).get('content', {}).pop(guid)
+            except KeyError:
+                pass
+            self.save()
+        finally:
+            self._lock.release()
+
+    def set_content_item_rebuild_status(self, server, guid, status):
+        """
+        Set the current status for a content rebuild
+        """
+        self._lock.acquire()
+        try:
+            content = self.get_content_item(server, guid)
+            content['rsconnect_rebuild_status'] = str(status)
+            self.save()
+        finally:
+            self._lock.release()
+
+    def get_content_items(self, server, status=None):
+        """
+        Get all the content items that are tracked for rebuild on the given server.
+        :param status: Filter results by rebuild status
+        :return: A list of content items
+        """
+        if status:
+            return [item for item in self._data.get(server.url, {}).get('content', {}).values() if item['rsconnect_rebuild_status'] == status]
+        else:
+            return list(self._data.get(server.url, {}).get('content', {}).values())
```

Comment:
nit: could factor out `self._data.get(server.url, {}).get('content', {}).values()` to make it clearer that the difference in the 2 cases is filtering vs. not filtering the same list.